spec/fastsync/vdd/fastsyncL2.tla

```diff
@@ -0,0 +1,679 @@
+    (* Therefore, we model this with nondeterministic behavior as it could lead to peer removal, for both correct and faulty. *)
+    (* See scheduler.go                                                                                                       *)
+    (* https://github.com/tendermint/tendermint/blob/4298bbcc4e25be78e3c4f21979d6aa01aede6e87/blockchain/v2/scheduler.go#L335 *)
+    LET toRemovePeers == IF nonDetFlag THEN bPool.peerIds ELSE {} IN
+       
+    (*
+      Corresponds to logic for pruning a peer that is responsible for delivering block for the next height.
+      If a peer responsible for next height is correct, we don't remove it as logic is based on difference between pending time
+      and current time. In the synchronous system model, correct peer will always respond timely. 
+      See scheduler.go     
+      https://github.com/tendermint/tendermint/blob/4298bbcc4e25be78e3c4f21979d6aa01aede6e87/blockchain/v2/scheduler.go#L617    
+    *)
+    LET nextHeightPeer == bPool.pendingBlocks[bPool.height] IN
+    LET prunablePeers == 
+        IF /\ nextHeightPeer /= NilPeer
+           /\ nextHeightPeer \in FAULTY
+           /\ nonDetFlag 
+        THEN toRemovePeers \union {nextHeightPeer} 
+        ELSE toRemovePeers
+    IN
+    RemovePeers(prunablePeers, bPool)
+    
+
+\* Handle SyncTimeout. It models if progress has been made (height has increased) since the last SyncTimeout event.
+HandleSyncTimeout(bPool) ==    
+    [bPool EXCEPT 
+            !.syncedBlocks = bPool.height - bPool.syncHeight, 
+            !.syncHeight = bPool.height
+    ]      
+
+
+\* Handle control messages.
+HandleControlMsg(msg, bPool) == 
+    IF msg.type = "statusResponse" 
+    THEN HandleStatusResponse(msg, bPool)
+    ELSE IF msg.type = "addPeer" 
+         THEN AddPeer(msg.peerId, bPool)
+         ELSE IF msg.type = "removePeer"
+         THEN RemovePeer(msg.peerId, bPool)  
+         ELSE IF msg.type = "syncTimeout"
+              THEN HandleSyncTimeout(bPool)  
+              ELSE bPool 
+
+HandleResponse(msg, bPool) == 
+    IF msg = NoMsg 
+    THEN bPool
+    ELSE IF msg.type = "blockResponse" 
+         THEN HandleBlockResponse(msg, bPool) 
+         ELSE HandleControlMsg(msg, bPool)  
+    
+(*
+   At every node step we executed the following steps (atomically):
+    1) input message is consumed and the corresponding handler is called,
+    2) pruning logic is called
+    3) block execution is triggered (we try to execute block at next height)
+    4) a request to a peer is made (if possible) and 
+    5) we decide if termination condition is satisifed so we stop. 
+*)
+NodeStep ==
+   LET bPool == HandleResponse(inMsg, blockPool) IN
+   LET bp == TryPrunePeer(bPool) IN
+   LET nbPool == ExecuteBlocks(bp) IN
+   LET msgAndPool == CreateRequest(nbPool) IN 
+   LET nstate == GetState(msgAndPool.pool) IN
+   
```

Comment:
This seems a bit strange at first look. The fact that we execute on multiple events in a single step. I need to be thinking a bit more on what the implications are but what was the reason for this? Also in the code we don't have the `GetState` as shown here and many of the steps above can send the `finished` event (we don't have an explicit state in the code). I think it should be ok but again, not sure why the change.

Reply:
We actually handle single event in a single step. At least the way how events are defined in this spec. On the other side we do multiple actions as part of single step; state update as part of request handling, pruning, and request creation. I was also hesitating about doing it this way, but it is very natural doing this at the high level protocol specification. It is basically, consume message, do state update and send message.  I was checking frequency at which we do stuff at the code level and based on that decided what should be part of single atomic step. Processing blocks, scheduling and pruning have similar frequency (pruning is a bit higher), and this is the reason why I put them all here as part of atomic step. SyncTimeout is at much lower frequency and therefore. it is externalised. The reason why GetState is isolated is because state is outside pool, and it was more convenient defining its logic separately. It should semantically be equivalent to the code although there it is being called at several places.